### PR TITLE
[Validator] Add the `WordCount` constraint

### LINKED ISSUE #6813: [RFC] [Validator] adding a new WordCount constraint validator
Hi guys,

I was thinking of adding a new constraint validator similar to the `Length` constraint but for counting words in a string. Sometimes, when writing a text or a description, we're asked to write at least x words or at most x words.

Do you think it would be needed in the validator component?

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #6813
| License       | MIT

Doc PR: https://github.com/symfony/symfony-docs/pull/20047

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/WordCount.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+        if (null !== $min && $min < 0) {
+            throw new ConstraintDefinitionException(\sprintf('The "%s" constraint requires the min word count to be a positive integer or 0 if set.', __CLASS__));
+        }
+
+        if (null !== $max && $max <= 0) {
+            throw new ConstraintDefinitionException(\sprintf('The "%s" constraint requires the max word count to be a positive integer if set.', __CLASS__));
+        }
+
+        if (null !== $this->min && null !== $this->max && $this->min > $this->max) {
```
Comment (derrabus) [inline src/Symfony/Component/Validator/Constraints/WordCount.php]:
```suggestion
        if (null !== $min && null !== $max && $min > $max) {
```

File: src/Symfony/Component/Validator/Constraints/WordCountValidator.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        if (!class_exists(\IntlBreakIterator::class)) {
+            throw new \RuntimeException(\sprintf('The "%s" constraint requires the "intl" PHP extension.', __CLASS__));
+        }
+
+        if (!$constraint instanceof WordCount) {
+            throw new UnexpectedTypeException($constraint, WordCount::class);
+        }
+
+        if (null === $value) {
```
Comment (derrabus) [inline src/Symfony/Component/Validator/Constraints/WordCountValidator.php]: An empty string should pass as well.

File: src/Symfony/Component/Validator/Tests/Constraints/WordCountTest.php [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+use Symfony\Component\Validator\Exception\ConstraintDefinitionException;
+use Symfony\Component\Validator\Exception\MissingOptionsException;
+use Symfony\Component\Validator\Mapping\ClassMetadata;
+use Symfony\Component\Validator\Mapping\Loader\AttributeLoader;
+
+class WordCountTest extends TestCase
+{
+    public function testLocaleIsSet()
+    {
+        $encoding = new WordCount(min: 1, locale: 'en');
```
Comment (derrabus) [inline src/Symfony/Component/Validator/Tests/Constraints/WordCountTest.php]: Why do you store the constraint in a variable named "encoding"?

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Validator/Tests/Constraints/WordCountTest.php (hunk 3 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/Validator/Tests/Constraints/WordCountTest.php]: That's a mistake from mine, that's updated now

File: src/Symfony/Component/Validator/Constraints/WordCount.php (hunk 1 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/Validator/Constraints/WordCount.php]: I made the update to be consistent with the rest of the constructor. Just curious, is there a reason beside consistency to do this change, or is that what you had in mind?

File: src/Symfony/Component/Validator/Tests/Constraints/WordCountValidatorTest.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    public function testNonStringValues(mixed $value)
+    {
+        $this->expectException(UnexpectedValueException::class);
+        $this->expectExceptionMessageMatches('/Expected argument of type "string", ".*" given/');
+
+        $this->validator->validate($value, new WordCount(min: 1));
+    }
+
+    public static function provideValidValues()
```
Comment (OskarStark) [inline src/Symfony/Component/Validator/Tests/Constraints/WordCountValidatorTest.php:77]:
```suggestion
    public static function provideValidValues(): iterable
```

File: src/Symfony/Component/Validator/Tests/Constraints/WordCountValidatorTest.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        yield ['my ascii string', 3];
+        yield ["   with a\nnewline", 3];
+        yield ["皆さん、こんにちは。", 4];
+        yield ["你好，世界！这是一个测试。", 9];
+        yield [new StringableValue('my ûtf 8'), 3];
+        yield [null, 1]; // null should always pass and eventually be handled by NotNullValidator
+        yield ['', 1]; // empty string should always pass and eventually be handled by NotBlankValidator
+    }
+
+    public static function provideInvalidTypes()
```
Comment (OskarStark) [inline src/Symfony/Component/Validator/Tests/Constraints/WordCountValidatorTest.php:88]:
```suggestion
    public static function provideInvalidTypes(): iterable
```

File: src/Symfony/Component/Validator/Tests/Constraints/WordCountValidatorTest.php (hunk 4 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/Validator/Tests/Constraints/WordCountValidatorTest.php:77]: Should we? Nicolas seemed to disagree on this point, hence the choice in this PR: https://github.com/symfony/symfony/pull/53154#discussion_r1436499093

Comment (OskarStark) [reply src/Symfony/Component/Validator/Tests/Constraints/WordCountValidatorTest.php:77]: Oh ok, I was only aware of all the `void` noise. I personally add iterable to my providers and also describe it via PHPDoc 🤷‍♂️

File: src/Symfony/Component/Validator/Constraints/WordCount.php (hunk 1 above)
Comment (derrabus) [reply src/Symfony/Component/Validator/Constraints/WordCount.php]: Yeah, consistency mainly. Also semantics: We're technically validating the constructor parameters, not the properties.

**Comment by alexandre-daubois** [comment]
Pushed plural form of the constraint message

File: src/Symfony/Component/Validator/Constraints/WordCountValidator.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        $words = iterator_to_array($iterator->getPartsIterator());
+
+        // erase "blank words" and don't count them as words
+        $wordsCount = \count(array_filter(array_map(trim(...), $words)));
+
+        if (null !== $constraint->min && $wordsCount < $constraint->min) {
+            $this->context->buildViolation($constraint->minMessage)
+                ->setParameter('{{ count }}', $wordsCount)
+                ->setParameter('{{ min }}', $constraint->min)
+                ->setPlural((int) $constraint->min)
```
Comment (AlexisRossat) [inline src/Symfony/Component/Validator/Constraints/WordCountValidator.php]: Is this cast needed because `$constraint->min` is already an int ?

Comment (alexandre-daubois) [reply src/Symfony/Component/Validator/Constraints/WordCountValidator.php:53]: Yes, `$min` is nullable. Casting `null` to int will result on 0, which is needed by `setPlural()` that only accepts `int`.

Comment (alexandre-daubois) [reply src/Symfony/Component/Validator/Constraints/WordCountValidator.php]: Nullability is checked above indeed, I unwrapped the useless cast :+1:

File: src/Symfony/Component/Validator/Constraints/WordCount.php [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        self::TOO_SHORT_ERROR => 'TOO_SHORT_ERROR',
+        self::TOO_LONG_ERROR => 'TOO_LONG_ERROR',
+    ];
+
+    #[HasNamedArguments]
+    public function __construct(
+        public ?int $min = null,
+        public ?int $max = null,
+        public ?string $locale = null,
+        public string $minMessage = 'This value is too short. It should contain at least {{ min }} word.|This value is too short. It should contain at least {{ min }} words.',
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/WordCount.php]:
```suggestion
        public string $minMessage = 'This value is too short. It should contain at least one word.|This value is too short. It should contain at least {{ min }} words.',
```

File: src/Symfony/Component/Validator/Constraints/WordCount.php [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        public string $maxMessage = 'This value is too long. It should contain {{ max }} word or less.|This value is too long. It should contain {{ max }} words or less.',
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/WordCount.php]:
```suggestion
        public string $maxMessage = 'This value is too long. It should contain one word or less.|This value is too long. It should contain {{ max }} words or less.',
```

File: src/Symfony/Component/Validator/Constraints/WordCount.php (hunk 7 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/Validator/Constraints/WordCount.php]: Here you go :+1:

File: src/Symfony/Component/Validator/Constraints/WordCount.php [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        self::TOO_LONG_ERROR => 'TOO_LONG_ERROR',
+    ];
+
+    #[HasNamedArguments]
+    public function __construct(
+        public ?int $min = null,
+        public ?int $max = null,
+        public ?string $locale = null,
+        public string $minMessage = 'This value is too short. It should contain at least one word.|This value is too short. It should contain at least {{ min }} words.',
+        public string $maxMessage = 'This value is too long. It should contain one word or less.|This value is too long. It should contain {{ max }} words or less.',
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/WordCount.php]: Thinking about this again: Is `1` actually a value that makes sense or shouldn't we enforce a value of `2` or higher?

In any case, if we were to keep `1` I propose a small rewording:

```suggestion
        public string $maxMessage = 'This value is too long. It should contain one word.|This value is too long. It should contain {{ max }} words or less.',
```

Comment (alexandre-daubois) [reply src/Symfony/Component/Validator/Constraints/WordCount.php]: It makes sense to me. I can easily imagine someone having an input where only one word is accepted (like a tag input for instance?). Also "one word" strongly depends on your locale. And it doesn't cost anything to accept this value so I'd say, let's authorize it?

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Component/Validator/Constraints/WordCount.php (hunk 9 above)
Comment (xabbuh) [reply src/Symfony/Component/Validator/Constraints/WordCount.php]: works for me 👍

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/Validator/Tests/Constraints/WordCountValidatorTest.php (hunk 4 above)
Comment (derrabus) [reply src/Symfony/Component/Validator/Tests/Constraints/WordCountValidatorTest.php:77]: Yeah, it's a Nicolas thing. 🙂

**Review by derrabus — APPROVED** [review]

**Review by xabbuh — CHANGES_REQUESTED** [review]
the failures on AppVeyor look related, we need to skip some of the added tests when the intl extension is missing

**Comment by alexandre-daubois** [comment]
Thought of adding a check in the constraint but forgot the tests. Thanks for pointing out, I added `@requires`

**Comment by OskarStark** [comment]
Can you please create and reference a docs PR? Thanks

**Review by xabbuh — APPROVED** [review]

**Comment by alexandre-daubois** [comment]
@OskarStark of course, here you go: https://github.com/symfony/symfony-docs/pull/20047

**Comment by xabbuh** [comment]
Thank you @alexandre-daubois.

**Pull request merged**